docs/RFCS/20220818_tenant_version_upgrades.md

```diff
@@ -0,0 +1,475 @@
+- [2020: Long running migrations](20200513_long_running_migrations.md)
+
+# Motivation
+
+We need to do work in this area because while we *knew* we needed to
+maintain invariants (as described above) in multi-tenancy, we actually
+failed to implement the code to check these invariants.
+
+In particular:
+
+- it's trivially possible to violate invariant D, because we do not implement
+  [the version interlock described above](#enforcing-the-invariants-during-upgrades-and-migrations-in-single-tenancy) in secondary tenants.
+
+  So different SQL pods running at different binary versions can
+  observe different logical versions temporarily, and expose different
+  feature sets (or even use system tables incorrectly). The TLV can
+  even appear to move backward.
+
+  We do not have a tracking issue for this yet.  A subset of the
+  problem is covered in this issue
+  https://github.com/cockroachdb/cockroach/issues/66606 .
+
+- we are missing a guardrail against violations of invariants E, F and
+  G: nothing blocks TLV upgrades beyond the SLV currently.
+
+  This is tracked here: https://github.com/cockroachdb/cockroach/issues/80992
+
+As a result of this lack of guardrails, it's possible to bring a
+multitenant cluster in an invalid state, where tenants are at a
+logical version beyond the level of support they need from the KV
+layer.
+
+This can cause serious UX pain and, in extreme case, outright data
+corruption.
+
+## Why we need to do this even in a shared-process world
+
+In [this
+proposal](https://github.com/cockroachdb/cockroach/pull/84700) we are
+thinking about running multiple tenants inside the same process,
+possibly shared with the KV layer.
+
+Does this simplify?
+
+Alas, it does not: it only forces the TBV to remain equal to SBV, but
+it does not constraint the TLV to remain "behind" the SLV.
+
+Additionally, it's also possible for different nodes (running
+different SQL servers for the "app" tenant) to run at different TBVs, and
+the SQL layer for secondary tenants does not properly persist its TLV, so
+it's still possible for the TLV to briefly appear to move backward.
+
+So invariants D, E, F and G can still be violated.
+
+So we still need guardrails with a shared-process deployment.
+
+# Technical design
+
+## SQL server startup guardrails
+
+We are going to block a SQL server from starting if its TBV is too low for the current TLV.
+
+We will do this with a RPC on startup that will check what is the current TLV for the tenant.
+
+## Version ratcheting guardrails
+
+We are going to extend [the
+interlock](#enforcing-the-invariants-during-upgrades-and-migrations-in-single-tenancy)
+previously implemented for single-tenancy to also work in secondary
+tenants:
+
+- there will be a new "migrate" function for secondary tenants.
+- at the beginning of each TLV step, it will check (with a RPC) that
+  the storage cluster (SLV) is already at the same logical version or newer.
```

Comment:
nit/optimization: every SQL server should probably just have a copy of the SLV for the storage cluster to which it is connected; we had to check it during SQL server boot anyway, so we can just store it; since we know it can only ever move up, we only need to use an RPC to refresh that copy if it is lower than the TLV we are going to; In practice, I'd expect this means we can do N steps of the TLV with 0 or 1 SLV check RPCs, since it is likely the SLV has moved all the way though its steps from major to major before the TLV moves at all.